File: demo_kennardstonealgorithm.py
# -*- coding: utf-8 -*- %reset -f
"""
@author: Hiromasa Kaneko
Demonstration of Kennard-Stone algorighm
"""

import matplotlib.pyplot as plt
import numpy as np

import kennardstonealgorithm

number_of_samples = 50
number_of_selected_samples = 20

# generate samples 0f samples for demonstration
X = np.random.rand(number_of_samples, 2)

# standarize X
autoscaledX = (X - X.mean()) / X.std(ddof=1)

# select samples using Kennard-Stone algorithm
selected_sample_numbers, remaining_sample_numbers = kennardstonealgorithm.kennardstonealgorithm(
    autoscaledX, number_of_selected_samples)
print("selected sample numbers")
print(selected_sample_numbers)
print("---")
print("remaining sample numbers")
print(remaining_sample_numbers)

# plot samples
plt.figure()
plt.scatter(autoscaledX[:, 0], autoscaledX[:, 1], label="all samples")
plt.scatter(autoscaledX[selected_sample_numbers, 0], autoscaledX[selected_sample_numbers, 1], marker="*",
            label="all samples")
plt.xlabel("x1")
plt.ylabel("x2")
plt.legend(loc='upper right')
plt.show()

File: demo_kennardstonealgorithm.m
% demonstration of Kennard-Stone algorighm

clear; close all;

numberofsamples = 50;
numberofselectedsamples = 20;

rng default

% generate samples of X-variables for demonstration
X = rand( numberofsamples, 2 );

% select samples using Kennard-Stone algorithm
[ selectedsamplenumbers, remainingsamplenumbers] = kennardstonealgorithm( X, numberofselectedsamples);
disp( 'selected sample numbers' );
disp( selectedsamplenumbers );
disp( '---' );
disp( 'remaining sample numbers' );
disp( remainingsamplenumbers );

% plot samples 
figure;
plot( X(:,1), X(:,2), 'b.', 'markersize', 10 );
hold on;
plot( X(selectedsamplenumbers,1), X(selectedsamplenumbers,2), 'ro', 'markersize', 8 );
hold off;
ylabel( 'x_2' ,  'FontSize' , 20 , 'FontName','Meiryo UI');
xlabel( 'x_1' ,  'FontSize' , 20 , 'FontName','Meiryo UI');
set(gcf, 'Color' , 'w' ); 
set(gca, 'FontSize' , 20);
set(gca, 'FontName','Meiryo UI');
% legend('all samples', 'selected samples','Location', 'northoutside');
legend('all samples', 'selected samples');
